### ISSUE TITLE
[Task 18][QianCyrus] Format-aware reward router

### ISSUE THREAD

**Issue by QianCyrus** [redai-infra/Relax#101]
## Task

Relax Contributor Program 2026 Cohort 1 — Task 18: Format-aware reward router.

Claim record: https://github.com/redai-infra/Relax/issues/86#issuecomment-5077510261

## 1. Repository, baseline commit, and development branch

- Repository: `redai-infra/Relax`
- Baseline branch: `main`
- Baseline commit: `039ce876d25540adad847d4223b4de4722d8f425`
- Development branch: `feat/format-aware-reward-router`
- Delivery: GitHub Issue + Draft Pull Request

The implementation and PR will be limited to Task 18. No unrelated formatting, algorithm, scheduler, or performance changes are in scope.

### CPU/unit-test environment

- OS: Red Hat Enterprise Linux 9.4
- Python: 3.12.13
- PyTest: 9.1.1
- PyTorch: 2.11.0+cu130
- CPU architecture: ARM64/aarch64
- GPU: not required by the router unit tests

### End-to-end validation environment

- Environment type: isolated Mamba environment built from scratch; no container image
- OS: Red Hat Enterprise Linux 9.4
- Python: 3.12
- GPU: 1 × NVIDIA GH200, 97,871 MiB visible memory
- Driver: 595.58.03
- CUDA toolkit: 13.2
- PyTorch: 2.11.0+cu130
- Ray: 2.56.1
- SGLang: 0.5.12.post1
- Transformer Engine: 2.14.1
- cuDNN: 9.19
- cuDNN frontend: 1.26
- Host architecture: ARM64/aarch64
- Parallelism: single-node, single-GPU colocate

The exact CPU model and host RAM were not part of the E2E acceptance snapshot and will not be used as a correctness metric. GPU identity, software versions, source commit, data hash, command, step count, and raw training log will be recorded.

### Targeted tests

```bash
python -m pytest -v \
  tests/engine/rewards/test_reward_router.py \
  tests/engine/rewards/test_reward_worker.py \
  tests/utils/test_arguments_opd_teacher_colocate.py
```

### Reward regression tests

```bash
python -m pytest -v \
  tests/engine/rewards/ \
  tests/utils/test_arguments_opd_teacher_colocate.py
```

### Repository checks

```bash
pre-commit run --all-files
```

### Mixed-reward 10-update E2E configuration

```bash
export RM_TYPE=
export RM_TYPE_FALLBACK=error
export NUM_ROLLOUT=10
export ROLLOUT_BATCH_SIZE=4
export N_SAMPLES=4
export GLOBAL_BATCH_SIZE=16
export ROLLOUT_MAX_RESPONSE_LEN=4096
export MAX_TOKENS_PER_GPU=4480
export LOGPROBS_MAX_TOKENS_PER_GPU=4480
export SGLANG_MEM_FRACTION_STATIC=0.35
export ATTENTION_BACKEND=auto

bash examples/beginner-task/run-qwen3-0.6B-1xgpu-grpo-mixed-reward.sh
```

Update arithmetic:

```text
10 rollouts × 4 prompts × 4 samples / global batch 16
= 10 optimizer updates
```

The E2E dataset contains 20 GSM8K prompts and 20 MMLU prompts. Each shuffled rollout batch contains two math prompts and two multiple-choice prompts.

### Planned files

- `relax/engine/rewards/registry.py`
- `relax/engine/rewards/__init__.py`
- `relax/utils/arguments.py`
- `tests/engine/rewards/test_reward_router.py`
- `tests/engine/rewards/test_reward_worker.py`
- `tests/utils/test_arguments_opd_teacher_colocate.py`
- `docs/en/guide/customize-training.md`
- `docs/zh/guide/customize-training.md`

### Planned design

1. Add a `RewardSpec` containing the reward name, sync/async mode, handler, and optional label matcher.
2. Add a `RewardRegistry` as the single source of reward registrations.
3. Add a pure `resolve_reward_route()` function that returns a structured route decision without Ray or GPU dependencies.
4. Preserve explicit `--rm-type` semantics for backward compatibility.
5. Add `--rm-type-fallback` with three policies:
   - `zero`: return zero and emit a warning;
   - `error`: emit a warning and raise a descriptive error;
   - a registered reward name: dispatch to that reward.
6. Detect unknown, missing, ambiguous, and metadata/label-conflicting routes.
7. Dispatch registered synchronous rewards through the existing worker pool and await registered asynchronous rewards directly.
8. Keep `boxed_` behavior compatible.
9. Deduplicate repeated warnings by route failure identity.

Routing precedence:

```text
metadata.rm_type
  → explicit --rm-type compatibility path
  → exactly one registered label matcher
  → configured --rm-type-fallback
```

The explicit CLI route remains ahead of automatic label inference so an existing single-task command keeps its previous behavior.

### Non-goals

This task will not:

- change math or multiple-choice scoring formulas;
- modify GRPO/PPO/DAPO algorithms;
- modify Controller, rollout scheduling, or weight synchronization;
- add third-party dependencies;
- change public `Sample.label` or `Sample.metadata` structures;
- introduce a general natural-language task classifier;
- modify existing datasets in bulk;
- claim performance or convergence improvements.

## 5. Baseline data, behavior, metrics, and targets

At baseline commit `039ce876d25540adad847d4223b4de4722d8f425`:

- `metadata.rm_type` can override the global `--rm-type`;
- dispatch is duplicated across a worker `if/elif` chain and async/sync type collections;
- no unified registry exists;
- no label-based route inference exists;
- no configurable fallback policy exists;
- no metadata/label conflict detection exists;
- an unknown type reaches a Ray worker and fails late with `RewardWorker: unknown rm_type=...`;
- a missing type fails with `Rule-based RM type is not specified.`;
- a valid but incorrect metadata type can silently invoke the wrong scorer.

Acceptance targets:

| Requirement | Measurement | Target |
|---|---|---|
| Mixed batch | Math + multiple-choice samples in one batch | Every sample reaches the expected reward |
| Metadata route | Valid `metadata.rm_type` | Dispatches to that registered reward |
| CLI compatibility | Existing explicit `--rm-type` | Existing route remains unchanged |
| Label route | No metadata or CLI, exactly one matcher | Dispatches to the matching reward |
| Conflict handling | Metadata disagrees with a recognized label | Configured fallback with `conflicting_reward_types` warning |
| Unknown/missing handling | Invalid or absent type | Configured zero/error/reward fallback |
| Registry extensibility | Register a test reward | One registration, no generic router branch change |
| Sync/async handling | Registered handlers of both modes | Correct execution path |
| Unit tests | Reward router/worker/argument tests | All pass |
| Code checks | `pre-commit run --all-files` | All pass |
| E2E integration | 10 rollouts and 10 optimizer updates | Complete, finite, no fallback/OOM/Traceback |

### Unit-level verification

- registry lookup and deterministic ordering;
- duplicate, empty, and invalid registrations;
- conservative math and multiple-choice label matchers;
- metadata, CLI, and label priority;
- mixed math + multiple-choice batch;
- unknown, invalid, missing, ambiguous, and conflicting routes;
- `zero`, `error`, and registered reward fallbacks;
- sync/async dispatch;
- `boxed_` compatibility;
- warning deduplication;
- legacy explicit `--rm-type` behavior.

### End-to-end verification

Public datasets:

- `openai/gsm8k`, revision `740312add88f781978c0658806c59bc2815b9866`, first 20 training rows;
- `cais/mmlu`, revision `c30699e8356da336a370243923dbaf21066bb9fe`, first 20 `high_school_computer_science` test rows.

Prepared JSONL SHA256:

```text
1a5a7bcf836a2e35f1f208ba741929464fb10b0c19fbc4bd5df3bfef6c78efe0
```

The E2E run will use no global `--rm-type` and `--rm-type-fallback error`. Therefore a missing, unknown, or conflicting sample route fails the run instead of being hidden by a zero reward.

Expected E2E counts:

```text
40 prompts
160 generated responses
80 math reward evaluations
80 multiple-choice reward evaluations
10 completed rollouts
10 unique optimizer updates, steps 0..9
```

E2E acceptance also requires finite policy loss, KL, and gradient norm; no OOM, mixed CUDA runtime, actor failure, or traceback; and a final validator bound to the unique run identity.

## 7. Expected delivery

The Draft PR will include:

- the reward registry and pure router;
- metadata/CLI/label routing;
- configurable fallback behavior and warnings;
- sync/async registry dispatch;
- backward compatibility tests;
- mixed-batch, conflict, fallback, and registry tests;
- English and Chinese documentation;
- reproducible test commands;
- a sanitized 10-step E2E result summary and relevant log excerpts.
